src/components/structures/ScrollPanel.js

```diff
@@ -51,7 +52,15 @@ module.exports = React.createClass({
 
         /* onFillRequest(backwards): a callback which is called on scroll when
          * the user nears the start (backwards = true) or end (backwards =
-         * false) of the list
+         * false) of the list.
+         *
+         * This should return true if the pagination was successful, or false if
+         * there is no more data in this directon (at this time) - which will
+         * stop the pagination cycle until the user scrolls again.
+         *
+         * This can return a promise; if it does, no more calls will be made
+         * until the promise completes. The promise should resolve to true or
+         * false as above.
```

Comment:
I'm not convinced it's a significant problem - 'either an X or a promise of an X' seems like a natural idiom and one which the q library makes it easy to handle.

Nevertheless I've changed the contract to require a promise, and changed the implementations of onFillRequest accordingly.

Done.